src/c_api/adapter.h

```diff
@@ -0,0 +1,275 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file adapter.h
+ */
+#ifndef XGBOOST_C_API_ADAPTER_H_
+#define XGBOOST_C_API_ADAPTER_H_
+#include <string>
+#include <limits>
+namespace xgboost {
+
+/** \brief External data formats should implement an adapter as below. The
+ * adapter provides a uniform access to data outside xgboost, allowing
+ * construction of DMatrix objects from a range of sources without duplicating
+ * code. The adapter should translate external data into batches of COO tuples
+ * containing the floating point value, row index and column index.*/
+class ExternalDataAdapter {
+ public:
+  ExternalDataAdapter(size_t num_features, size_t num_rows, size_t num_elements)
+      : num_features(num_features),
+        num_rows(num_rows),
+        num_elements(num_elements) {}
+  struct COOTuple {
+    COOTuple(size_t row_idx, size_t column_idx, float value)
+        : row_idx(row_idx), column_idx(column_idx), value(value) {}
+
+    size_t row_idx{0};
+    size_t column_idx{0};
+    float value{0};
+  };
+  size_t GetNumFeatures() const { return num_features; }
+  size_t GetNumRows() const { return num_rows; }
+  size_t GetNumElements() const { return num_elements; }
+
+ protected:
+  size_t num_features;
+  size_t num_rows;
+  size_t num_elements;
+};
+
+class CSRAdapter : public ExternalDataAdapter {
+ public:
+  CSRAdapter(const size_t* row_ptr, const unsigned* feature_idx,
+             const float* values, size_t num_rows, size_t num_elements,
+             size_t num_features)
+      : ExternalDataAdapter(num_features, num_rows, num_elements),
+        row_ptr(row_ptr),
+        feature_idx(feature_idx),
+        values(values) {}
+
+ private:
+  class CSRAdapterBatch {
+   public:
+    CSRAdapterBatch(size_t row_idx, size_t size, const unsigned* feature_idx,
+                    const float* values)
+        : row_idx(row_idx),
+          size(size),
+          feature_idx(feature_idx),
+          values(values) {}
+
+    size_t Size() const { return size; }
+    COOTuple GetElement(size_t idx) const {
+      return COOTuple(row_idx, feature_idx[idx], values[idx]);
+    }
+
+   private:
+    size_t row_idx;
+    size_t size;
+    const unsigned* feature_idx;
+    const float* values;
+  };
+
+ public:
+  size_t Size() const { return num_rows; }
+  const CSRAdapterBatch operator[](size_t idx) const {
```

Comment:
Could you write a `\brief` for what is the semantic meaning of `operator[]`, first glance it wasn't clear to me what exactly is a batch for different adapters.